### PR TITLE
fix: improve the providing of discriminator models

### PR DESCRIPTION
**Pull request by lafeuil**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

## What is the current behavior?
When I declare some discriminators and I start my project, I get the following errors : 
```
OverwriteModelError: Cannot overwrite `XYG` model once compiled
```
I think this bug is due to the MR #660 that declares submodel twice : 
1) with the `model` function
2) with the `discriminator` function 

In the source code, there is a comment and I think this MR fix it :
> discriminators must convert to `AsyncModelFactory`.
> Otherwise, the discriminators will register as `Model` before `model.discriminator` and throw OverwriteModelError
> 

I only reproduce the bug when I start my project and I can't reproduce it with unit test. 

Issue Number: #658, #660

## What is the new behavior?
This MR use only `discriminator` function from the parent model to declare submodel. The submodel are provided in nestjs context by declaring a provider and the parent model is injected in the `useFactory` function to call `discriminator` function

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

**Comment by kamilmysliwiec** [comment]
Can you please create an issue with a minimum reproduction repository (which shows the error you're facing atm).

**Comment by hanselabreu** [comment]
I tried this locally and the error still persists...

File: lib/mongoose.providers.ts [hunk 1]
```diff
@@ -1,20 +1,7 @@
 import { Provider } from '@nestjs/common';
 import { Connection, Document, Model } from 'mongoose';
 import { getConnectionToken, getModelToken } from './common/mongoose.utils';
-import {
-  AsyncModelFactory,
-  DiscriminatorOptions,
-  ModelDefinition,
-} from './interfaces';
-
-function addDiscriminators(
```
Comment (hanselabreu) [inline lib/mongoose.providers.ts:10]: could probably keep this function, just change it to: 
```
function addDiscriminators(
  discriminators: DiscriminatorOptions[] = [],
  option: ModelDefinition | AsyncModelFactory,
) {
  return discriminators.map((d) => ({
    provide: getModelToken(d.name),
    useFactory: (model: Model<Document>) =>
      model.discriminator(d.name, d.schema),
    inject: [getModelToken(option.name)],
  }));
}
```
and then just spread it's return `...addDiscriminators(option.discriminators, option)`

**Review by hanselabreu — APPROVED** [review]

**Comment by johnfrades** [comment]
hopefully this gets merged soon as im also experiencing the same problem with Discriminators

**Comment by dpolicastro** [comment]
Any news on this?

**Comment by jmcdo29** [comment]
@kamilmysliwiec issue #697 has a reproduction. [I'll link it here as well](https://github.com/lafeuil/nestjs-mongoose-test).

I know there are several people who have ran into this.

**Comment by leif-eric93** [comment]
Also experiencing the issue

**Comment by dpolicastro** [comment]
For those looking for a workaround to make it work until this fix is merged, you can find it here https://github.com/nestjs/mongoose/issues/697#issuecomment-752246688

**Comment by kamilmysliwiec** [comment]
LGTM

**Pull request merged**